xsnap/documentation/Compartments.md

```diff
@@ -0,0 +1,618 @@
+
+	import * as foo from "./moduleMap-fixture.js";
+	
+	let getterCount = 0;
+	let setterCount = 0;
+	const moduleMap = {
+		foo,
+		get bar() {
+			getterCount++;
+			return this.foo;
+		},
+		set bar(it) {
+			setterCount++;
+			this.foo = it;
+		},
+	};
+	
+	const c1 = new Compartment({}, moduleMap, {});
+	const fooNS1 = await c1.import("foo");
+	const barNS1 = await c1.import("bar");
+	
+	const c2 = new Compartment({}, moduleMap, {});
+	const fooNS2 = await c2.import("foo");
+	const barNS2 = await c2.import("bar");
+	
+	print(getterCount); // 2;
+	print(setterCount); // 0;
+	print(fooNS1.default()); // 0;
+	print(barNS1.default()); // 1;
+	print(fooNS2.default()); // 2;
+	print(barNS2.default()); // 3;
+
+#### module descriptors
+
+When a module map property value is a module descriptor, the module descriptor is used to load and initialize a new module into the compartment.
+
+	const foo = { source:`
+		let foo = 0;
+		export default function() {
+			return foo++;
+		}
+	`};
+	
+	let getterCount = 0;
+	let setterCount = 0;
+	const moduleMap = {
+		foo,
+		get bar() {
+			getterCount++;
+			return this.foo;
+		},
+		set bar(it) {
+			setterCount++;
+			this.foo = it;
+		},
+	};
+	
+	const c1 = new Compartment({}, moduleMap, {});
+	const fooNS1 = await c1.import("foo");
+	const barNS1 = await c1.import("bar");
+	
+	const c2 = new Compartment({}, moduleMap, {});
+	const fooNS2 = await c2.import("foo");
+	const barNS2 = await c2.import("bar");
+	
+	print(getterCount); // 2;
+	print(setterCount); // 0;
+	print(fooNS1.default()); // 0;
+	print(barNS1.default()); // 0;
+	print(fooNS2.default()); // 0;
+	print(barNS2.default()); // 0;
+
+#### module specifiers
+
+When a module map property value is a module specifier, the module specifier is used by the parent compartment to load and initialize a new module into the compartment.
+
+	const uri = import.meta.uri;
+	const foo = uri.slice(0, uri.lastIndexOf("/") + 1) + "moduleMap-fixture.js";
+	
+	let getterCount = 0;
+	let setterCount = 0;
+	const moduleMap = {
+		foo,
+		get bar() {
+			getterCount++;
+			return this.foo;
+		},
+		set bar(it) {
+			setterCount++;
+			this.foo = it;
+		},
+	};
+	
+	const c1 = new Compartment({}, moduleMap, {});
+	const fooNS1 = await c1.import("foo");
+	const barNS1 = await c1.import("bar");
+	
+	const c2 = new Compartment({}, moduleMap, {});
+	const fooNS2 = await c2.import("foo");
+	const barNS2 = await c2.import("bar");
+	
+	print(getterCount); // 2;
+	print(setterCount); // 0;
+	print(fooNS1.default()); // 0;
+	print(barNS1.default()); // 0;
+	print(fooNS2.default()); // 0;
+	print(barNS2.default()); // 0;
+
+
+### options.globalLexicals
+
+The `globalLexicals` option is an object that initializes the global lexical scope of the compartment:
+
+- property names become variable or constant names,
+- property values become variable or constant values. 
+
+If the property is writable, a variable (`let`) is created, else a constant (`const`) is created.
```

Comment:
Are only enumerable properties of `globalLexicals` captured exactly once and before the Compartment constructor returns?

I would expect this test to pass:
```js
const globalLexicals = {};
const compartment = new Compartment({}, {}, {globalLexicals});
globalLexicals.foo = 10;
assert(compartment.evaluate(`typeof foo === 'undefined'`));
```

Reply:
The behavior documented here seems consistent with our intentions. There is a separate question of a `globalContour` which would, when enabled, only apply to evaluated code and persist between calls to `evaluate`, such that a constant introduced in the top-level scope of the given script would be introduced to the global contour. The global contour would be an append only lexical scope. We have not attempted to produce a feature like this in the SES shim. XS might be in a better position to experiment with a global contour design. The motivating use case would be a REPL.

I believe the globalContour would have to overshadow globalLexicals if those features composed, making the stack look like [globalThis, globalLexicals, globalContour]. Declarations in the evaluated program would exist in the globalContour scope.

Consequently, declarations of the evaluated script should _not_ collide with names in the global lexical scope. I do not know whether that’s the behavior implemented here, so another test might be useful.

```js
const c = new Compartment({}, {}, {globalLexicals: Object.freeze({foo: {}})});
c.evaluate(`const foo = 10`); // should not throw
c.evaluate(`const foo = 10`); // should also not throw
```

Assuming the existence of a globalContour feature, enabled by a `globalContourEnabled` flag in compartment options:

```js
const c = new Compartment({}, {}, {globalContourEnabled: true});
c.evaluate(`const foo = 10`);
c.evaluate(`const foo = 20`); // should throw
```